src/txn/txn_rollback_to_stable.c

```diff
@@ -137,6 +71,190 @@ __txn_abort_newer_insert(
+
+    /* Serialize the update. */
+    WT_ERR(__wt_update_serial(session, page, upd_entry, &upd, upd_size, true));
+
+err:
+    return (ret);
+}
+
+/*
+ * __txn_abort_row_replace_with_hs_value --
+ *     Abort updates in the history store and replace the on-disk with an update that satisfy than
+ *     the given timestamp.
+ */
+static int
+__txn_abort_row_replace_with_hs_value(
+  WT_SESSION_IMPL *session, WT_PAGE *page, WT_ROW *rip, wt_timestamp_t rollback_timestamp)
+{
+    WT_CURSOR *hs_cursor;
+    WT_CURSOR_BTREE *cbt;
+    WT_DECL_ITEM(hs_key);
+    WT_DECL_ITEM(hs_value);
+    WT_DECL_ITEM(key);
+    WT_DECL_RET;
+    WT_TIME_PAIR hs_start, hs_stop;
+    WT_UPDATE *hs_upd, *upd;
+    wt_timestamp_t durable_ts;
+    size_t size;
+    uint32_t hs_btree_id, session_flags;
+    uint8_t prepare_state, type;
+    int cmp;
+    bool valid_update_found;
+
+    hs_cursor = NULL;
+    hs_upd = upd = NULL;
+    hs_btree_id = S2BT(session)->id;
+    session_flags = 0;
+    valid_update_found = false;
+
+    /* Allocate buffers for the data store and history store key. */
+    WT_RET(__wt_scr_alloc(session, 0, &key));
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_key));
+    WT_ERR(__wt_scr_alloc(session, 0, &hs_value));
+
+    WT_ERR(__wt_row_leaf_key(session, page, rip, key, false));
+
+    /* Open a history store table cursor. */
+    WT_ERR(__wt_hs_cursor(session, &session_flags));
+    hs_cursor = session->hs_cursor;
+    cbt = (WT_CURSOR_BTREE *)hs_cursor;
+
+    /*
+     * Scan the history store for the given btree and key with maximum start and stop time pair to
+     * let the search point to the last version of the key and start traversing backwards to find
+     * out the satisfying record according the given timestamp. Any satisfying history store record
+     * is moved into data store and remove it from history store. If none of the history store
+     * records satisfy the given timestamp, the key is removed from data store.
+     */
+    ret = __wt_hs_cursor_position(session, hs_cursor, hs_btree_id, key, WT_TS_MAX);
+    for (; ret == 0; ret = hs_cursor->prev(hs_cursor)) {
+        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, hs_key, &hs_start.timestamp,
+          &hs_start.txnid, &hs_stop.timestamp, &hs_stop.txnid));
+
+        /* Stop before crossing over to the next btree */
+        if (hs_btree_id != S2BT(session)->id)
+            break;
+
+        /*
+         * Keys are sorted in an order, skip the ones before the desired key, and bail out if we
+         * have crossed over the desired key and not found the record we are looking for.
+         */
+        WT_ERR(__wt_compare(session, NULL, hs_key, key, &cmp));
+        if (cmp != 0)
+            break;
+
+        /* Set this comparison as exact match of the search for later use. */
+        cbt->compare = 0;
+
+        /* Stop processing when we find the update that is less than the given update. */
+        if (hs_start.timestamp <= rollback_timestamp) {
```

Comment:
we have to compare the durable timestamp as well. Only if the durable timestamp is less than rollback_timestamp we can stop else we need to continue.

Reply:
Used the durable timestamp.